magma/primitives/slice.py

```diff
@@ -40,15 +42,32 @@ def slice(*args, **kwargs):
     return get_slice(*args, **kwargs)
 
 
-def set_slice(target: Bits, value: Bits, start: UInt, width: int):
+def set_slice(target: Bits, value: Bits, start: Union[UInt, int], width: int):
     """
     target: the output with a dynamic slice range replaced by value
     value: the output driving a slice of target
     start: dynamic start index of the slice
     width: constant slice width
     """
-    if not isinstance(start, UInt):
-        raise TypeError("start should be a UInt")
+    if not isinstance(start, (UInt, int)):
+        raise TypeError("start should be a UInt or int")
+    if len(target) == 1:
+        if start != 0:
+            raise ValueError(
+                "Start must equal 0 for set_slice with value of length 1")
+        if width != 1:
+            raise ValueError(
+                "Width must equal 1 for set_slice with value of length 1")
+        if not (
+            isinstance(value, type(target)) or
+            isinstance(target, type(value))
+        ):
+            raise ValueError(
+                "Incompatible types found for target and value")
+        return value
```

Comment:
We could allow a non constant value, but that would be "unsafe" if start was ever non zero